<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Linkage">
<title>Monogenic MOTIF analysis • Linkage</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Monogenic MOTIF analysis">
<meta property="og:description" content="Linkage">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-slim navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Linkage</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/Install.html">Install</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-vignettes">Vignettes</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-vignettes">
    <h6 class="dropdown-header" data-toc-skip>R package</h6>
    <a class="dropdown-item" href="../articles/LinkageR.html">Genes and regulatory peaks to find transcription factors</a>
    <a class="dropdown-item" href="../articles/Motif.html">Monogenic MOTIF analysis</a>
    <a class="dropdown-item" href="../articles/Enrichment.html">Pathway Enrichment Module</a>
    <h6 class="dropdown-header" data-toc-skip>Shiny APP</h6>
    <a class="dropdown-item" href="../articles/Linkage.html">Linkage in shiny APP</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/aicplane/Linkage/issues">FAQ</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="http://www.xulabgdpu.org.cn/linkage">Shiny App</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://aicplane.github.io/Linkage/">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/aicplane/Linkage">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Monogenic MOTIF analysis</h1>
            
      
      
      <div class="d-none name"><code>Motif.Rmd</code></div>
    </div>

    
    
<div class="section level3">
<h3 id="create-linkage-object">
<strong>1.Create Linkage Object</strong><a class="anchor" aria-label="anchor" href="#create-linkage-object"></a>
</h3>
<div align="justify">
<p>The chromatin accessibility data is a normalized bulk ATAC-seq count
matrix, which a prior count of 5 is added to the raw counts, then put
into a “counts per million”, then log2 transformed, then quantile
normalized. The gene expression data is a normalized bulk RNA-seq count
matrix, which is log2(fpkm+1) transformed. Two tab-delimited text/csv
input files (chromatin accessibility matrix and gene expression matrix)
are required before running Linkage. The gene expression matrix file is
a tab-delimited multi-column data matrix, which the first column
represents gene symbols and the following columns represent normalized
or raw expression levels of genes for each sample. The chromatin
accessibility matrix file is a tab-delimited multi-column data matrix as
well, which the first three columns represent chromosome name, start
coordinate on the chromosome and end coordinate on the chromosome of the
peaks respectively; the remaining columns of the chromatin accessibility
matrix file represent normalized or raw chromatin accessibility levels
of peaks for each sample. <strong>Each element of ready-to-analysis
files will appear in Chromatin Accessibility Matrix panel and Gene
Expression Matrix panel.</strong></p>
</div>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://aicplane.github.io/Linkage/" class="external-link">Linkage</a></span><span class="op">)</span></span>
<span><span class="va">RNA.seq.dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"TCGA-BRCA-RNA.txt"</span>, package <span class="op">=</span> <span class="st">"Linkage"</span><span class="op">)</span></span>
<span><span class="va">RNA.seq</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="va">RNA.seq.dir</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">Homo.list.files.dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Homo.ATAC"</span>, package <span class="op">=</span> <span class="st">"Linkage"</span><span class="op">)</span></span>
<span><span class="va">Homo.list.files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">Homo.list.files.dir</span><span class="op">)</span></span>
<span><span class="va">Homo.list.files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">Homo.list.files.dir</span>, <span class="st">"/"</span>, <span class="va">Homo.list.files</span><span class="op">)</span></span>
<span><span class="va">Homo.df_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">Homo.list.files</span>, <span class="kw">function</span><span class="op">(</span><span class="va">file</span><span class="op">)</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="va">file</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ATAC.seq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">Homo.df_list</span><span class="op">)</span></span>
<span><span class="va">LinkageObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CreateLinkageObject.html">CreateLinkageObject</a></span><span class="op">(</span>ATAC_count <span class="op">=</span> <span class="va">ATAC.seq</span>, RNA_count <span class="op">=</span> <span class="va">RNA.seq</span>, Species <span class="op">=</span> <span class="st">"Homo"</span>, id_type <span class="op">=</span> <span class="st">"ensembl_gene_id"</span><span class="op">)</span></span>
<span><span class="va">LinkageObject</span></span></code></pre></div>
<pre><code><span><span class="co">## An LinkageObject </span></span>
<span><span class="co">## 59161 gene 215920 peak </span></span>
<span><span class="co">## Active gene: 0 </span></span>
<span><span class="co">## Active peak: positive peak 0 negetive peak 0 </span></span>
<span><span class="co">## Active TF: 0</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="regulatory-peaks-search">
<strong>2.Regulatory Peaks Search</strong><a class="anchor" aria-label="anchor" href="#regulatory-peaks-search"></a>
</h3>
<div align="justify">
<p><strong>Regulatory Peaks Search</strong> allows users to detect all
potential regulatory DNA regions for specific genes. When given an
<strong>input gene</strong> and <strong>targeting DNA regions</strong>,
Linkage will automatically perform <strong>canonical correlation
analysis</strong> between each quantitative chromatin accessibility
measure in the region and the quantitative expression level of the gene
across all samples. Users can easily adjust the region of interest and
customize the correlation algorithm (<strong>Spearman</strong> /
<strong>Pearson</strong> / <strong>Kendall</strong>). Then, all the
statistically significant results are listed in the Potential
Cis-regulatory Regions panel. With clicking on a specific region of this
panel, users can view the scatter plot of quantitative chromatin
accessibility and gene expression from the Correlation Plot panel. The
<strong>rho</strong> and <strong>FDR</strong> for correlation analysis
also be shown on the scatter plot.</p>
</div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TSPAN6"</span>, <span class="st">"CD99"</span>, <span class="st">"KLHL13"</span><span class="op">)</span></span>
<span><span class="va">LinkageObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/regulatory_peak.html">regulatory_peak</a></span><span class="op">(</span>LinkageObject <span class="op">=</span> <span class="va">LinkageObject</span>, gene_list <span class="op">=</span> <span class="va">gene_list</span>, genelist_idtype <span class="op">=</span> <span class="st">"external_gene_name"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="motif-analysis-module">3.Motif Analysis Module<a class="anchor" aria-label="anchor" href="#motif-analysis-module"></a>
</h3>
<div align="justify">
<p>Motif Analysis Module supports users to visualize the enriched
transcription factor motifs within potential regulatory peaks. With
clicking on a specific peak, users can view the location and binding
score information of each enriched transcription factor of this region.
Once users select one of the transcription factors, the corresponding
sequence logo of this cis-regulatory element will appear in the
Sequence-logo Plot panel.</p>
</div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">motif</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SingleMotifAnalysis.html">SingleMotifAnalysis</a></span><span class="op">(</span><span class="va">LinkageObject</span><span class="op">@</span><span class="va">cor.peak</span><span class="op">$</span><span class="va">TSPAN6</span>, <span class="st">"Homo"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/seqLogo_plot.html">seqLogo_plot</a></span><span class="op">(</span><span class="st">"MA0618.1"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Motif_files/figure-html/unnamed-chunk-4-1.png" width="672"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Zenghui Liu.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
